### PR TITLE
Adding TensorFlow benchmarks

### PR DESCRIPTION
**Pull request by ma2rten**

### CODE REVIEW

**Comment by vrv** [comment]
FYI, I am getting different results when I run using the latest nightly (cuda 7.5, cudnnv4, turned off ondemand cpu governor [1], Intel(R) Xeon(R) CPU E5-1650 v3 @ 3.50GHz, Titan X):

(Note that I have corrected this to -s 30):

$ python rnn.py -n rnn -b 20 -i 100 -l 100 -s 30 2>/dev/null
Forward:
--- 100000 samples in 10.9685809612 seconds (9116.948627 samples/s) ---
Forward + Backward:
--- 100000 samples in 27.2731859684 seconds (3666.604999 samples/s) ---
$ python rnn.py -n rnn -b 20 -i 100 -l 500 -s 30 2>/dev/null
Forward:
--- 100000 samples in 21.2142031193 seconds (4713.822855 samples/s) ---
Forward + Backward:
--- 100000 samples in 42.3182210922 seconds (2363.048237 samples/s) ---
$ python rnn.py -n basic_lstm -b 20 -i 100 -l 100 -s 30 2>/dev/null
Forward:
--- 100000 samples in 30.3018350601 seconds (3300.130167 samples/s) ---
Forward + Backward:
--- 100000 samples in 85.1440079212 seconds (1174.480757 samples/s) ---
$ python rnn.py -n basic_lstm -b 20 -i 100 -l 500 -s 30 2>/dev/null
Forward:
--- 100000 samples in 37.6162240505 seconds (2658.427314 samples/s) ---
Forward + Backward:
--- 100000 samples in 92.8219828606 seconds (1077.330990 samples/s) ---
$ python rnn.py -n lstm -b 20 -i 100 -l 100 -s 30 2>/dev/null
Forward:
--- 100000 samples in 29.2092580795 seconds (3423.571841 samples/s) ---
Forward + Backward:
--- 100000 samples in 86.6548669338 seconds (1154.003260 samples/s) ---
$ python rnn.py -n lstm -b 20 -i 100 -l 500 -s 30 2>/dev/null
Forward:
--- 100000 samples in 36.4524109364 seconds (2743.302683 samples/s) ---
Forward + Backward:
--- 100000 samples in 94.4019358158 seconds (1059.300287 samples/s) ---

Thus:

#### Hidden layer size 100 - Batch size 20

| Library | Time (µs) | Forward only (µs) |
| --- | --- | --- |
| TensorFlow | 272.7 | 109.7 |

#### Hidden layer size 500 - Batch size 20

| Library | Time (µs) | Forward only (µs) |
| --- | --- | --- |
| TensorFlow | 423.1 | 212.1 |

### LSTM

#### Hidden layer size 100 - Batch size 20

| Library | Time (µs) | Forward only (µs) |
| --- | --- | --- |
| TensorFlow (LSTM) | 866.5 | 292.1 |
| TensorFlow (BasicLSTM) | 851.4 | 303.0 |

#### Hidden layer size 500 - Batch size 20

| Library | Time (µs) | Forward only (µs) |
| --- | --- | --- |
| TensorFlow (LSTM) | 944.0 | 364.5 |
| TensorFlow (BasicLSTM) | 928.2 | 376.2 |

I will leave some other comments.  These benchmarks (for all systems) should consistently perform timing correctly, should ignore a few warm-up runs in the timing, etc.  @soumith has done a great job maintaining this consistency at https://github.com/soumith/convnet-benchmarks

FYI @ebrevdo @zheng-xq

[1] Turning on performance governor:
    sudo bash -c 'for i in `ls /sys/devices/system/cpu/*/cpufreq/scaling_governor`; do echo 'performance' > $i; done;'

File: tensorflow/rnn.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+optim = tf.train.GradientDescentOptimizer(0.01)
+train_op = optim.minimize(cost)
+
+session = tf.Session()
+session.run(tf.initialize_all_variables())
+
+start = time.time()
+for k, i in enumerate(xrange(0, n_samples, batch_size)):
+    session.run(output[-1], feed_dict=get_feed_dict(X_data[i:i+batch_size])) 
+
```
Comment (vrv) [inline tensorflow/rnn.py:64]: You probably want to set end = time.time() here instead of calling it several times on line 66, same below.

File: tensorflow/rnn.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+cost = tf.reduce_mean((output[-1] - y) ** 2)
+
+optim = tf.train.GradientDescentOptimizer(0.01)
+train_op = optim.minimize(cost)
+
+session = tf.Session()
+session.run(tf.initialize_all_variables())
+
+start = time.time()
+for k, i in enumerate(xrange(0, n_samples, batch_size)):
```
Comment (vrv) [inline tensorflow/rnn.py:62]: It is generally a good idea (in all frameworks) to run a few 'warmup steps' outside of the timing loop to make sure you measure steady-state performance instead of startup performance (which could be measured separately).

**Comment by glample** [comment]
Thanks for the benchmarks!
I'm not able to reproduce them on my machine though (not even half the time). I installed tensorflow following these steps:
https://www.tensorflow.org/versions/r0.7/get_started/os_setup.html#pip-installation

Everything looks fine when I run the script:
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcurand.so locally
I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 0 with properties: 
name: GeForce GTX TITAN X
major: 5 minor: 2 memoryClockRate (GHz) 1.076
pciBusID 0000:02:00.0
Total memory: 12.00GiB
Free memory: 7.29GiB
I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 1 with properties: 
name: GeForce GTX TITAN X
major: 5 minor: 2 memoryClockRate (GHz) 1.076
pciBusID 0000:84:00.0
Total memory: 12.00GiB
Free memory: 8.92GiB

When I run the code for hidden layer 100, GPU usage never goes above 25%.
Am I supposed to do anything special to optimize the library?

**Comment by vrv** [comment]
Hm, I didn't have to do anything -- I just ran the script as is.  I also only have one Titan X -- what happens if you try to only make one visible to the process?  (set CUDA_VISIBLE_DEVICES=0 as one way of doing this, though there are others).

https://github.com/tensorflow/tensorflow#installation you can also try installing a nightly pip package to see if that changes anything.

When you say you only get 25% GPU usage, does that mean you're getting results, but not the same numbers as the ones I reported?

**Comment by glample** [comment]
Yes, I obtain these times:
RNN (hidden layer 100): 689.6 | 265.1
RNN (hidden layer 500): 862.1 | 316.1
LSTM (hidden layer 100): 2087.9 | 737.2
LSTM (hidden layer 500): 2171.7 | 777.0
TensorFlow on my machine never uses more than 30% of the GPU, when Torch uses basically 100%

**Comment by vrv** [comment]
I would try my suggestions in the last post to narrow down the discrepancies - I had a freshly installed Ubuntu and the pip nightly from last night when I produced those results.

**Comment by glample** [comment]
I tried the pip nightly as well as showing only one GPU, but it's not faster.
Although nothing is running on my GPU, there are 5/12GB used by some other processes. Can it make a difference?
Could you maybe try to run quickly one of the torch or theano scripts? I would be curious to see how fast they are on your machine.

**Comment by vrv** [comment]
Theano (not an expert user), installed from pip today.

$ THEANO_FLAGS=floatX=float32,device=gpu0  python rnn.py -i 100 -s 30 -l 100 -b 20
Using gpu device 0: GeForce GTX TITAN X (CNMeM is disabled, CuDNN not available)
compiling...
Setup : compile + forward/backward x 1
--- 76.0730199814 seconds
Forward:
--- 100000 samples in 6.69198203087 seconds (14943.255905 samples/s, 0.0000669 s/sample) ---
Forward + Backward:
--- 100000 samples in 17.6506638527 seconds (5665.509288 samples/s, 0.0001765 s/sample) ---

So my numbers are even better than the ones you reported, though not by the same magnitude as the difference between our TF runs.

File: tensorflow/rnn.py (hunk 1 above)
Comment (ma2rten) [inline tensorflow/rnn.py:64]: I agree, but this code is copied from the theano benchmark. Should be changed there too.

**Comment by ma2rten** [comment]
The reason that our benchmarks is likely that I changed from reduce_mean to reduce_sum, otherwise we are using the exact same setup.

**Comment by vrv** [comment]
TF now supports reduce_mean for GPU, so that likely isn't the problem.  We're looking into the source of the discrepancies, since we're all getting different results on the same GPU -- it points to something CPU related (even though I've verified all ops are running on GPU).

**Comment by ma2rten** [comment]
Actually I didn't turn off cpu governor and I am not sure if I used nightly or stable, I will rerun again once there is nothing else running on that server.

**Comment by ma2rten** [comment]
My` updated timing using Tensorflow nightly, cuda 7.5, cudnnv4, cpu governor turned off, Intel(R) Xeon(R) CPU E5-1650 v3 @ 3.50GHz, Titan X:

```
+ python rnn.py -n rnn -b 20 -i 100 -l 100 -s 30
Forward:
 --- 100000 samples in 11.2845079899 seconds (8861.706695 samples/s) ---
Forward + Backward:
--- 100000 samples in 25.0517599583 seconds (3991.735328 samples/s) ---
+ python rnn.py -n rnn -b 20 -i 100 -l 500 -s 30
Forward:
--- 100000 samples in 19.6468820572 seconds (5089.865906 samples/s) ---
Forward + Backward:
--- 100000 samples in 41.4640939236 seconds (2411.725139 samples/s) ---
+ python rnn.py -n basic_lstm -b 20 -i 100 -l 100 -s 30
Forward:
--- 100000 samples in 26.9969060421 seconds (3704.128033 samples/s) ---
Forward + Backward:
--- 100000 samples in 76.8980400562 seconds (1300.423245 samples/s) ---
+ python rnn.py -n basic_lstm -b 20 -i 100 -l 500 -s 30
Forward:
--- 100000 samples in 37.4096050262 seconds (2673.110150 samples/s) ---
Forward + Backward:
--- 100000 samples in 101.383234024 seconds (986.356371 samples/s) ---
+ python rnn.py -n lstm -b 20 -i 100 -l 100 -s 30
Forward:
--- 100000 samples in 26.6797909737 seconds (3748.155001 samples/s) ---
Forward + Backward:
--- 100000 samples in 79.9265930653 seconds (1251.148007 samples/s) ---
+ python rnn.py -n lstm -b 20 -i 100 -l 500 -s 30
Forward:
--- 100000 samples in 36.8016619682 seconds (2717.268516 samples/s) ---
Forward + Backward:
--- 100000 samples in 103.259371996 seconds (968.435089 samples/s) ---
```

**Comment by glample** [comment]
I tried on another machine, I don't really have better results. However on my GPU machines, CPUs are pretty slow (E5-2620 v3 @ 2.40GHz). Can it be the reason why I don't obtain the same results?

**Comment by vrv** [comment]
Probably -- it looks like for these small LSTM cells there's some thread contention and CPU overhead -- the GPU isn't really busy in most of these benchmarks for any framework.

I should also note that I think these benchmarks aren't benchmarking the same computation -- I believe the Theano numbers are actually benchmarking smaller matrix multiplies than the others, at least in one or two cases. For example: https://github.com/glample/rnn-benchmarks/blob/master/theano/rnn.py#L50 vs. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell.py#L122

It would be prudent to carefully validate that the benchmarks in all frameworks are computing the same overall function.
